src/main/kotlin/org/elm/lang/core/psi/ElmPsiManager.kt

```diff
@@ -7,53 +7,73 @@
+     *
+     * It is incremented whenever a non-whitespace, non-comment change is made to the PSI of an Elm
+     * file outside of function bodies.
+     */
     val modificationTracker = SimpleModificationTracker()
 
     override fun projectOpened() {
         PsiManager.getInstance(project).addPsiTreeChangeListener(CacheInvalidator())
     }
 
     inner class CacheInvalidator : PsiTreeChangeAdapter() {
-        override fun beforeChildRemoval(event: PsiTreeChangeEvent) = onPsiChange(event)
-        override fun childReplaced(event: PsiTreeChangeEvent) = onPsiChange(event)
-        override fun childAdded(event: PsiTreeChangeEvent) = onPsiChange(event)
-        override fun childrenChanged(event: PsiTreeChangeEvent) = onPsiChange(event)
-        override fun childMoved(event: PsiTreeChangeEvent) = onPsiChange(event)
-        override fun propertyChanged(event: PsiTreeChangeEvent) = onPsiChange(event)
-
-        private fun onPsiChange(event: PsiTreeChangeEvent) {
+        override fun beforeChildRemoval(event: PsiTreeChangeEvent) = onPsiChange(event, event.child)
+        override fun beforeChildReplacement(event: PsiTreeChangeEvent) = onPsiChange(event, event.oldChild)
+        override fun beforeChildMovement(event: PsiTreeChangeEvent) = onPsiChange(event, event.child)
+        override fun childReplaced(event: PsiTreeChangeEvent) = onPsiChange(event, event.newChild)
+        override fun childAdded(event: PsiTreeChangeEvent) = onPsiChange(event, event.child)
+        override fun childMoved(event: PsiTreeChangeEvent) = onPsiChange(event, event.child)
+        override fun childrenChanged(event: PsiTreeChangeEvent) {
             // `GenericChange` event means that "something changed in the file" and sends
             // after all events for concrete PSI changes in a file.
-            // We handle more concrete events and so should ignore generic event
-            if (event is PsiTreeChangeEventImpl && event.isGenericChange) return
+            // We handle more concrete events and so should ignore the generic event.
+            if (event !is PsiTreeChangeEventImpl || !event.isGenericChange) onPsiChange(event, event.parent)
+        }
 
-            // There are some cases when PsiFile stored in the event as a child
-            // e.g. file removal by external VFS change
-            val file = event.file ?: event.child as? PsiFile
-            if (file?.fileType != ElmFileType) return
+        override fun propertyChanged(event: PsiTreeChangeEvent) {
+            if (event.propertyName != PROP_WRITABLE && event.element != null) {
+                onPsiChange(event, event.element)
+            }
+        }
+
+        private fun onPsiChange(event: PsiTreeChangeEvent, element: PsiElement) {
+            // If the file is null, then this is an event about VFS changes
+            val file = event.file
+            if (file == null && (element is ElmFile || element is PsiDirectory)) {
```

Comment:
If we have an easy way to check that a `PsiDirectory` is in an Elm source root, we could add that here as an optimization, but I don't think it's a big deal.

Reply:
`org.elm.workspace.ElmWorkspaceService#findProjectForFile`. But I agree, it's probably not necessary.